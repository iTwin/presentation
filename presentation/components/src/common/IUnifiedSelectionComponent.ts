/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Unified Selection */

import { IModelConnection } from "@bentley/imodeljs-frontend";
import { SelectionHandler } from "@bentley/presentation-frontend";

export default interface IUnifiedSelectionComponent {
  imodel: IModelConnection;
  rulesetId: string;
  selectionHandler?: SelectionHandler;
}
