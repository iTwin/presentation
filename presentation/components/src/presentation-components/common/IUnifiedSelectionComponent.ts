/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Core
 */

import { SelectionHandler } from "@bentley/presentation-frontend";
import { IModelConnection } from "@bentley/imodeljs-frontend";

/**
 * An interface for all unified selection components
 * @public
 */
export interface IUnifiedSelectionComponent {
  /** [[IModelConnection]] used by this data provider */
  readonly imodel: IModelConnection;

  /** Selection handler used by this component */
  readonly selectionHandler?: SelectionHandler;
}
