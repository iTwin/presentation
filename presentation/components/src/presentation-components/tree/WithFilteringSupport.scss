/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

.filteredTree {

  position: relative;
  width: 100%;
  height: 100%;

  .filteredTreeOverlay {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 9999;
    background-color: rgba(255,255,255,.8);
  }

}
