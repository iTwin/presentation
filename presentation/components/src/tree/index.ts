/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

export { default as PresentationTreeDataProvider } from "./DataProvider";
export { default as IPresentationTreeDataProvider } from "./IPresentationTreeDataProvider";
export { default as withUnifiedSelection, SelectionTarget } from "./WithUnifiedSelection";
export { default as withFilteringSupport } from "./WithFilteringSupport";
