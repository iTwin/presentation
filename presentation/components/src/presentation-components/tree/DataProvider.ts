/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Tree
 */

import memoize from "micro-memoize";
import { IDisposable, Logger } from "@bentley/bentleyjs-core";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { HierarchyRequestOptions, NodeKey, NodePathElement, Ruleset } from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";
import { DelayLoadedTreeNodeItem, PageOptions, TreeNodeItem } from "@bentley/ui-components";
import { RulesetRegistrationHelper } from "../common/RulesetRegistrationHelper";
import { IPresentationTreeDataProvider } from "./IPresentationTreeDataProvider";
import { createTreeNodeItems, pageOptionsUiToPresentation, PRESENTATION_TREE_NODE_KEY } from "./Utils";

/**
 * Properties for creating a `PresentationTreeDataProvider` instance.
 * @public
 */
export interface PresentationTreeDataProviderProps {
  /** IModel to pull data from. */
  imodel: IModelConnection;
  /** Id of the ruleset to use when requesting content or a ruleset itself. */
  ruleset: string | Ruleset;
  /**
   * Paging size for obtaining nodes.
   *
   * Presentation data providers, when used with paging, have ability to save one backend request for size / count. That
   * can only be achieved when `pagingSize` property is set on the data provider and it's value matches size which is used when
   * requesting nodes. To help developers notice this problem, data provider emits a warning similar to this:
   * ```
   * PresentationTreeDataProvider.pagingSize doesn't match pageOptions in PresentationTreeDataProvider.getNodes call. Make sure you set PresentationTreeDataProvider.pagingSize to avoid excessive backend requests.
   * ```
   * To fix the issue, developers should make sure the page size used for requesting data is also set for the data provider:
   * ```TS
   * const pagingSize = 10;
   * const provider = new TreeDataProvider({imodel, ruleset, pagingSize});
   * // only one backend request is made for the two following requests:
   * provider.getNodesCount();
   * provider.getNodes({ start: 0, size: pagingSize });
   * ```
   */
  pagingSize?: number;
}

/**
 * Presentation Rules-driven tree data provider.
 * @public
 */
export class PresentationTreeDataProvider implements IPresentationTreeDataProvider, IDisposable {
  private _imodel: IModelConnection;
  private _rulesetRegistration: RulesetRegistrationHelper;
  private _pagingSize?: number;

  /** Constructor. */
  public constructor(props: PresentationTreeDataProviderProps) {
    this._rulesetRegistration = new RulesetRegistrationHelper(props.ruleset);
    this._imodel = props.imodel;
    this._pagingSize = props.pagingSize;
  }

  /** Destructor. Must be called to clean up.  */
  public dispose() {
    this._rulesetRegistration.dispose();
  }

  /** Id of the ruleset used by this data provider */
  public get rulesetId(): string { return this._rulesetRegistration.rulesetId; }

  /** [[IModelConnection]] used by this data provider */
  public get imodel(): IModelConnection { return this._imodel; }

  /**
   * Paging options for obtaining nodes.
   * @see `PresentationTreeDataProviderProps.pagingSize`
   */
  public get pagingSize(): number | undefined { return this._pagingSize; }
  public set pagingSize(value: number | undefined) { this._pagingSize = value; }

  /** Called to get extended options for node requests */
  private createRequestOptions(): HierarchyRequestOptions<IModelConnection> {
    return {
      imodel: this._imodel,
      rulesetOrId: this._rulesetRegistration.rulesetId,
    };
  }

  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   * **Warning:** the `node` must be created by this data provider.
   */
  public getNodeKey(node: TreeNodeItem): NodeKey {
    return (node as any)[PRESENTATION_TREE_NODE_KEY];
  }

  /**
   * Returns nodes
   * @param parentNode The parent node to return children for.
   * @param pageOptions Information about the requested page of data.
   */
  public async getNodes(parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<DelayLoadedTreeNodeItem[]> {
    if (undefined !== pageOptions && pageOptions.size !== this.pagingSize) {
      const msg = `PresentationTreeDataProvider.pagingSize doesn't match pageOptions in PresentationTreeDataProvider.getNodes call.
        Make sure you set PresentationTreeDataProvider.pagingSize to avoid excessive backend requests.`;
      Logger.logWarning("Presentation.Components", msg);
    }

    return (await this._getNodesAndCount(parentNode, pageOptions)).nodes;
  }

  /**
   * Returns the total number of nodes
   * @param parentNode The parent node to return children count for.
   */
  public async getNodesCount(parentNode?: TreeNodeItem): Promise<number> {
    const pageOptions = undefined !== this.pagingSize ? { start: 0, size: this.pagingSize } : undefined;
    return (await this._getNodesAndCount(parentNode, pageOptions)).count;
  }

  private _getNodesAndCount = memoize(async (parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<{ nodes: TreeNodeItem[], count: number }> => {
    const requestCount = undefined !== pageOptions && 0 === pageOptions.start && undefined !== pageOptions.size;
    const parentKey = parentNode ? this.getNodeKey(parentNode) : undefined;

    if (!requestCount) {
      const allNodes = await Presentation.presentation.getNodes({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
      return { nodes: parentNode ? createTreeNodeItems(allNodes, parentNode.id) : createTreeNodeItems(allNodes), count: allNodes.length };
    }

    const nodesResponse = await Presentation.presentation.getNodesAndCount({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
    return { nodes: parentNode ? createTreeNodeItems(nodesResponse.nodes, parentNode.id) : createTreeNodeItems(nodesResponse.nodes), count: nodesResponse.count };
  }, { isMatchingKey: MemoizationHelpers.areNodesRequestsEqual as any });

  /**
   * Returns filtered node paths.
   * @param filter Filter.
   */
  public getFilteredNodePaths = async (filter: string): Promise<NodePathElement[]> => {
    return Presentation.presentation.getFilteredNodePaths(this.createRequestOptions(), filter);
  }

  /**
   * Loads the hierarchy so on-demand requests and filtering works quicker
   * @alpha Hierarchy loading performance needs to be improved before this becomes publicly available.
   */
  public async loadHierarchy() {
    return Presentation.presentation.loadHierarchy(this.createRequestOptions());
  }

}

class MemoizationHelpers {
  public static areNodesRequestsEqual(lhsArgs: [TreeNodeItem?, PageOptions?], rhsArgs: [TreeNodeItem?, PageOptions?]): boolean {
    if (lhsArgs[0]?.id !== rhsArgs[0]?.id)
      return false;
    if ((lhsArgs[1]?.start ?? 0) !== (rhsArgs[1]?.start ?? 0))
      return false;
    if ((lhsArgs[1]?.size ?? 0) !== (rhsArgs[1]?.size ?? 0))
      return false;
    return true;
  }
}
