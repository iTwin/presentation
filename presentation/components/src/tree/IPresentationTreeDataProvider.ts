/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

import { ITreeDataProvider, TreeNodeItem } from "@bentley/ui-components/lib/tree/TreeDataProvider";
import IPresentationDataProvider from "../common/IPresentationDataProvider";
import { NodeKey, NodePathElement } from "@bentley/presentation-common";

/**
 * Presentation tree data provider.
 */
export default interface IPresentationTreeDataProvider extends ITreeDataProvider, IPresentationDataProvider {
  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   */
  getNodeKey(node: TreeNodeItem): NodeKey;

  /**
   * Returns filtered node paths.
   */
  getFilteredNodePaths(filter: string): Promise<NodePathElement[]>;
}
