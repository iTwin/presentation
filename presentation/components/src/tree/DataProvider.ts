/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

import * as _ from "lodash";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { NodeKey, NodePathElement, HierarchyRequestOptions } from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";
import { DelayLoadedTreeNodeItem, TreeNodeItem } from "@bentley/ui-components/lib/tree/TreeDataProvider";
import { PageOptions } from "@bentley/ui-components/lib/common/PageOptions";
import { createTreeNodeItems, pageOptionsUiToPresentation } from "./Utils";
import IPresentationTreeDataProvider from "./IPresentationTreeDataProvider";

/**
 * Presentation Rules-driven tree data provider.
 */
export default class PresentationTreeDataProvider implements IPresentationTreeDataProvider {
  private _rulesetId: string;
  private _connection: IModelConnection;

  /**
   * Constructor.
   * @param connection Connection to an imodel to pull data from.
   * @param rulesetId Id of the ruleset used by this data provider.
   */
  public constructor(connection: IModelConnection, rulesetId: string) {
    this._rulesetId = rulesetId;
    this._connection = connection;
  }

  /** Id of the ruleset used by this data provider */
  public get rulesetId(): string { return this._rulesetId; }

  /** [[IModelConnection]] used by this data provider */
  public get connection(): IModelConnection { return this._connection; }

  /** Called to get extended options for node requests */
  private createRequestOptions(): HierarchyRequestOptions<IModelConnection> {
    return {
      imodel: this._connection,
      rulesetId: this._rulesetId,
    };
  }

  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   * **Warning:** the `node` must be created by this data provider.
   */
  public getNodeKey(node: TreeNodeItem): NodeKey {
    return node.extendedData.key as NodeKey;
  }

  /**
   * Returns nodes
   * @param parentNode The parent node to return children for.
   * @param pageOptions Information about the requested page of data.
   */
  public getNodes = _.memoize(async (parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<DelayLoadedTreeNodeItem[]> => {
    if (parentNode) {
      const parentKey = this.getNodeKey(parentNode);
      const childNodes = await Presentation.presentation.getChildren({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
      return createTreeNodeItems(childNodes, parentNode.id);
    }
    const rootNodes = await Presentation.presentation.getRootNodes({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) });
    return createTreeNodeItems(rootNodes);
  }, MemoizationHelpers.getNodesKeyResolver);

  /**
   * Returns the total number of nodes
   * @param parentNode The parent node to return children count for.
   */
  public getNodesCount = _.memoize(async (parentNode?: TreeNodeItem): Promise<number> => {
    if (parentNode) {
      const parentKey = this.getNodeKey(parentNode);
      return await Presentation.presentation.getChildrenCount(this.createRequestOptions(), parentKey);
    } else {
      return await Presentation.presentation.getRootNodesCount(this.createRequestOptions());
    }
  }, MemoizationHelpers.getNodesCountKeyResolver);

  /**
   * Returns filtered node paths.
   * @param filter Filter.
   */
  public getFilteredNodePaths = async (filter: string): Promise<NodePathElement[]> => {
    return Presentation.presentation.getFilteredNodePaths(this.createRequestOptions(), filter);
  }
}

class MemoizationHelpers {
  public static createKeyForPageOptions(pageOptions?: PageOptions) {
    if (!pageOptions)
      return "0/0";
    return `${(pageOptions.start) ? pageOptions.start : 0}/${(pageOptions.size) ? pageOptions.size : 0}`;
  }
  public static createKeyForTreeNodeItem(item?: TreeNodeItem) { return item ? item.id : ""; }
  public static getNodesKeyResolver(parent?: TreeNodeItem, pageOptions?: PageOptions) {
    return `${MemoizationHelpers.createKeyForTreeNodeItem(parent)}/${MemoizationHelpers.createKeyForPageOptions(pageOptions)}`;
  }
  public static getNodesCountKeyResolver(parent?: TreeNodeItem) { return MemoizationHelpers.createKeyForTreeNodeItem(parent); }
}
