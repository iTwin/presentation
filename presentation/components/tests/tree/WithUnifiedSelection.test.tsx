/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import "@bentley/presentation-frontend/tests/_helpers/MockFrontendEnvironment";
import * as React from "react";
import { expect, spy } from "chai";
import { mount, shallow } from "enzyme";
import * as faker from "faker";
import * as moq from "@bentley/presentation-common/tests/_helpers/Mocks";
import { createRandomECInstanceNodeKey } from "@bentley/presentation-common/tests/_helpers/random";
import { createRandomTreeNodeItem } from "../_helpers/UiComponents";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { KeySet, BaseNodeKey, ECInstanceNodeKey } from "@bentley/presentation-common";
import {
    Presentation,
    SelectionHandler, SelectionManager, SelectionChangeEvent, ISelectionProvider, SelectionChangeEventArgs, SelectionChangeType,
} from "@bentley/presentation-frontend";
import { Tree, TreeNodeItem, UiComponents } from "@bentley/ui-components";
import { I18N } from "@bentley/imodeljs-i18n";
import { DataTreeProps as TreeProps } from "@bentley/ui-components/lib/tree/component/DataTree";
import IUnifiedSelectionComponent from "../../lib/common/IUnifiedSelectionComponent";
import { IPresentationTreeDataProvider, withUnifiedSelection, SelectionTarget } from "../../lib/tree";

// tslint:disable-next-line:variable-name naming-convention
const PresentationTree = withUnifiedSelection(Tree);

describe("Tree withUnifiedSelection", () => {
    before(async () => {
        const i18n = new I18N([], "");
        await UiComponents.initialize(i18n);
    });

    let testRulesetId: string;
    const imodelMock = moq.Mock.ofType<IModelConnection>();
    const dataProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
    const selectionHandlerMock = moq.Mock.ofType<SelectionHandler>();
    beforeEach(() => {
        testRulesetId = faker.random.word();
        selectionHandlerMock.reset();
        setupDataProvider();
    });

    const setupDataProvider = (providerMock?: moq.IMock<IPresentationTreeDataProvider>, imodel?: IModelConnection, rulesetId?: string, rootNodes?: () => TreeNodeItem[], childNodes?: (parent: TreeNodeItem) => TreeNodeItem[]) => {
        if (!providerMock)
            providerMock = dataProviderMock;
        if (!imodel)
            imodel = imodelMock.object;
        if (!rulesetId)
            rulesetId = testRulesetId;
        if (!rootNodes)
            rootNodes = () => [];
        if (!childNodes)
            childNodes = () => [];
        providerMock.reset();
        providerMock.setup((x) => x.connection).returns(() => imodel!);
        providerMock.setup((x) => x.rulesetId).returns(() => rulesetId!);
        providerMock.setup((x) => x.onTreeNodeChanged).returns(() => undefined);
        providerMock.setup((x) => x.getNodeKey(moq.It.isAny())).returns((n: TreeNodeItem) => n.extendedData.key);
        providerMock.setup((x) => x.getRootNodes(moq.It.isAny())).returns(async () => rootNodes!());
        providerMock.setup((x) => x.getRootNodesCount()).returns(async () => rootNodes!().length);
        providerMock.setup((x) => x.getChildNodes(moq.It.isAny(), moq.It.isAny())).returns(async (p) => childNodes!(p));
        providerMock.setup((x) => x.getChildNodesCount(moq.It.isAny())).returns(async (p) => childNodes!(p).length);
    };

    it("mounts", () => {
        mount(<PresentationTree
            dataProvider={dataProviderMock.object}
            selectionHandler={selectionHandlerMock.object}
        />);
    });

    it("uses data provider's imodel and rulesetId", () => {
        const component = shallow(<PresentationTree
            dataProvider={dataProviderMock.object}
            selectionHandler={selectionHandlerMock.object}
        />).instance() as any as IUnifiedSelectionComponent;

        expect(component.imodel).to.equal(imodelMock.object);
        expect(component.rulesetId).to.equal(testRulesetId);
    });

    it("creates default implementation for selection handler when not provided through props", () => {
        const selectionManagerMock = moq.Mock.ofType<SelectionManager>();
        selectionManagerMock.setup((x) => x.selectionChange).returns(() => new SelectionChangeEvent());
        Presentation.selection = selectionManagerMock.object;

        const tree = shallow(<PresentationTree
            dataProvider={dataProviderMock.object}
        />).instance() as any as IUnifiedSelectionComponent;

        expect(tree.selectionHandler).to.not.be.undefined;
        expect(tree.selectionHandler!.name).to.not.be.undefined;
        expect(tree.selectionHandler!.rulesetId).to.eq(testRulesetId);
        expect(tree.selectionHandler!.imodel).to.eq(imodelMock.object);
    });

    it("renders correctly", () => {
        expect(shallow(<PresentationTree
            dataProvider={dataProviderMock.object}
            selectionHandler={selectionHandlerMock.object}
        />)).to.matchSnapshot();
    });

    it("disposes selection handler when unmounts", () => {
        const tree = shallow(<PresentationTree
            dataProvider={dataProviderMock.object}
            selectionHandler={selectionHandlerMock.object}
        />);
        tree.unmount();
        selectionHandlerMock.verify((x) => x.dispose(), moq.Times.once());
    });

    it("updates selection handler and data provider when props change", () => {
        const tree = shallow<TreeProps>(<PresentationTree
            dataProvider={dataProviderMock.object}
            selectionHandler={selectionHandlerMock.object}
        />);

        const imodelMock2 = moq.Mock.ofType<IModelConnection>();
        const rulesetId2 = faker.random.word();
        const providerMock2 = moq.Mock.ofType<IPresentationTreeDataProvider>();
        setupDataProvider(providerMock2, imodelMock2.object, rulesetId2);

        tree.setProps({
            dataProvider: providerMock2.object,
        });

        selectionHandlerMock.verify((x) => x.imodel = imodelMock2.object, moq.Times.once());
        selectionHandlerMock.verify((x) => x.rulesetId = rulesetId2, moq.Times.once());
    });

    it("handles missing selection handler when unmounts", () => {
        const component = shallow(<PresentationTree
            dataProvider={dataProviderMock.object}
            selectionHandler={selectionHandlerMock.object}
        />, { disableLifecycleMethods: true });
        component.unmount();
    });

    it("handles missing selection handler when updates", () => {
        const component = shallow(<PresentationTree
            dataProvider={dataProviderMock.object}
            selectionHandler={selectionHandlerMock.object}
        />, { disableLifecycleMethods: true });
        component.instance().componentDidUpdate!(component.props(), component.state()!);
    });

    describe("selection handling", () => {

        describe("checking if node should be selected", () => {

            it("returns true if node's id is in selectedNodes array prop", () => {
                const node = createRandomTreeNodeItem();
                const arr = [faker.random.uuid(), node.id, faker.random.uuid()];

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                    selectedNodes={arr}
                />);

                const propCallback = tree.find(Tree).prop("selectedNodes") as ((node: TreeNodeItem) => boolean);
                const actualResult = propCallback(node);

                expect(actualResult).to.eq(true);
            });

            it("returns false if node's id is not in selectedNodes array prop", () => {
                const node = createRandomTreeNodeItem();
                const arr = [faker.random.uuid(), faker.random.uuid()];

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                    selectedNodes={arr}
                />);

                const propCallback = tree.find(Tree).prop("selectedNodes") as ((node: TreeNodeItem) => boolean);
                const actualResult = propCallback(node);

                expect(actualResult).to.eq(false);
            });

            it("calls props callback and returns its result", () => {
                const node = createRandomTreeNodeItem();
                const result = faker.random.boolean();
                const callback = moq.Mock.ofType<(node: TreeNodeItem) => boolean>();
                callback.setup((x) => x(node)).returns(() => result).verifiable();

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                    selectedNodes={callback.object}
                />);

                const propCallback = tree.find(Tree).prop("selectedNodes") as ((node: TreeNodeItem) => boolean);
                const actualResult = propCallback(node);

                callback.verifyAll();
                expect(actualResult).to.eq(result);
            });

            it("returns false when there's no selection handler", () => {
                const node = createRandomTreeNodeItem();
                selectionHandlerMock.setup((x) => x.getSelection()).returns(() => new KeySet());

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                />, { disableLifecycleMethods: true });

                const propCallback = tree.find(Tree).prop("selectedNodes") as ((node: TreeNodeItem) => boolean);
                const result = propCallback(node);
                expect(result).to.be.false;
            });

            it("returns true when node key is in selection", () => {
                const nodeKey = createRandomECInstanceNodeKey();
                const node = createRandomTreeNodeItem();
                node.extendedData.key = nodeKey;
                selectionHandlerMock.setup((x) => x.getSelection()).returns(() => new KeySet([nodeKey]));

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                />);

                const propCallback = tree.find(Tree).prop("selectedNodes") as ((node: TreeNodeItem) => boolean);
                const result = propCallback(node);
                expect(result).to.be.true;
            });

            it("returns true when ECInstance key of ECInstance node is in selection", () => {
                const nodeKey = createRandomECInstanceNodeKey();
                const node = createRandomTreeNodeItem();
                node.extendedData.key = nodeKey;
                selectionHandlerMock.setup((x) => x.getSelection()).returns(() => new KeySet([nodeKey.instanceKey]));

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                />);

                const propCallback = tree.find(Tree).prop("selectedNodes") as ((node: TreeNodeItem) => boolean);
                const result = propCallback(node);
                expect(result).to.be.true;
            });

            it("returns false when node key is not in selection and node is not ECInstance node", () => {
                const nodeKey: BaseNodeKey = {
                    type: faker.random.word(),
                    pathFromRoot: [],
                };
                const node = createRandomTreeNodeItem();
                node.extendedData.key = nodeKey;
                selectionHandlerMock.setup((x) => x.getSelection()).returns(() => new KeySet());

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                />);

                const propCallback = tree.find(Tree).prop("selectedNodes") as ((node: TreeNodeItem) => boolean);
                const result = propCallback(node);
                expect(result).to.be.false;
            });

        });

        describe("selecting nodes", () => {

            it("calls props callback and adds node keys to selection manager when callback returns true", () => {
                const nodes = [createRandomTreeNodeItem(), createRandomTreeNodeItem()];
                const callback = moq.Mock.ofType<(nodes: TreeNodeItem[], replace: boolean) => boolean>();
                callback.setup((x) => x(nodes, false)).returns(() => true).verifiable();

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                    onNodesSelected={callback.object}
                />);

                tree.find(Tree).prop("onNodesSelected")!(nodes, false);

                selectionHandlerMock.verify((x) => x.addToSelection(nodes.map((n) => n.extendedData.key)), moq.Times.once());
                selectionHandlerMock.verify((x) => x.replaceSelection(moq.It.isAny()), moq.Times.never());
                callback.verifyAll();
            });

            it("calls props callback and aborts when it returns false", () => {
                const nodes = [createRandomTreeNodeItem(), createRandomTreeNodeItem()];
                const callback = moq.Mock.ofType<(nodes: TreeNodeItem[], replace: boolean) => boolean>();
                callback.setup((x) => x(nodes, true)).returns(() => false).verifiable();

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                    onNodesSelected={callback.object}
                />);

                tree.find(Tree).prop("onNodesSelected")!(nodes, true);

                selectionHandlerMock.verify((x) => x.addToSelection(moq.It.isAny()), moq.Times.never());
                selectionHandlerMock.verify((x) => x.replaceSelection(moq.It.isAny()), moq.Times.never());
                callback.verifyAll();
            });

            it("returns false when there's no selection handler", () => {
                const nodes = [createRandomTreeNodeItem()];
                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                />, { disableLifecycleMethods: true });

                tree.find(Tree).prop("onNodesSelected")!(nodes, true);

                selectionHandlerMock.verify((x) => x.addToSelection(moq.It.isAny()), moq.Times.never());
                selectionHandlerMock.verify((x) => x.replaceSelection(moq.It.isAny()), moq.Times.never());
            });

            it("replaces ECInstance keys in selection manager when selection target is set to `ECInstance`", () => {
                const nodes = [createRandomTreeNodeItem(), createRandomTreeNodeItem()];
                nodes[0].extendedData.key = createRandomECInstanceNodeKey();
                nodes[1].extendedData.key = { type: faker.random.word(), pathFromRoot: [] };

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                    selectionTarget={SelectionTarget.Instance}
                />);

                tree.find(Tree).prop("onNodesSelected")!(nodes, true);

                selectionHandlerMock.verify((x) => x.addToSelection(moq.It.isAny()), moq.Times.never());
                selectionHandlerMock.verify((x) => x.replaceSelection([
                    (nodes[0].extendedData.key as ECInstanceNodeKey).instanceKey,
                    nodes[1].extendedData.key,
                ]), moq.Times.once());
            });

        });

        describe("deselecting nodes", () => {

            it("calls props callback and removes node keys from selection manager when callback returns true", () => {
                const nodes = [createRandomTreeNodeItem(), createRandomTreeNodeItem()];
                const callback = moq.Mock.ofType<(nodes: TreeNodeItem[]) => boolean>();
                callback.setup((x) => x(nodes)).returns(() => true).verifiable();

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                    onNodesDeselected={callback.object}
                />);

                tree.find(Tree).prop("onNodesDeselected")!(nodes);

                selectionHandlerMock.verify((x) => x.removeFromSelection(nodes.map((n) => n.extendedData.key)), moq.Times.once());
                callback.verifyAll();
            });

            it("calls props callback and aborts when it returns false", () => {
                const nodes = [createRandomTreeNodeItem(), createRandomTreeNodeItem()];
                const callback = moq.Mock.ofType<(nodes: TreeNodeItem[]) => boolean>();
                callback.setup((x) => x(nodes)).returns(() => false).verifiable();

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                    onNodesDeselected={callback.object}
                />);

                tree.find(Tree).prop("onNodesDeselected")!(nodes);

                selectionHandlerMock.verify((x) => x.removeFromSelection(moq.It.isAny()), moq.Times.never());
                callback.verifyAll();
            });

            it("returns false when there's no selection handler", () => {
                const nodes = [createRandomTreeNodeItem()];
                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                />, { disableLifecycleMethods: true });

                tree.find(Tree).prop("onNodesDeselected")!(nodes);

                selectionHandlerMock.verify((x) => x.removeFromSelection(moq.It.isAny()), moq.Times.never());
            });

            it("removes ECInstance keys from selection manager when selection target is set to `ECInstance`", () => {
                const nodes = [createRandomTreeNodeItem(), createRandomTreeNodeItem()];
                nodes[0].extendedData.key = createRandomECInstanceNodeKey();
                nodes[1].extendedData.key = { type: faker.random.word(), pathFromRoot: [] };

                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                    selectionTarget={SelectionTarget.Instance}
                />);

                tree.find(Tree).prop("onNodesDeselected")!(nodes);

                selectionHandlerMock.verify((x) => x.removeFromSelection([
                    (nodes[0].extendedData.key as ECInstanceNodeKey).instanceKey,
                    nodes[1].extendedData.key,
                ]), moq.Times.once());
            });

        });

        describe("reacting to unified selection changes", () => {

            const triggerSelectionChange = (selectionLevel: number) => {
                const args: SelectionChangeEventArgs = {
                    changeType: SelectionChangeType.Clear,
                    imodel: imodelMock.object,
                    level: selectionLevel,
                    source: selectionHandlerMock.name,
                    timestamp: new Date(),
                    keys: new KeySet(),
                };
                const selectionProviderMock = moq.Mock.ofType<ISelectionProvider>();
                selectionProviderMock.setup((x) => x.getSelection(imodelMock.object, moq.It.isAny())).returns(() => new KeySet());
                selectionHandlerMock.target.onSelect!(args, selectionProviderMock.object);
            };

            it("re-renders tree on selection changes when selection level is 0", () => {
                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                />);
                const s = spy.on(tree.instance(), Tree.prototype.render.name);
                triggerSelectionChange(0);
                expect(s).to.be.called();
            });

            it("doesn't re-render tree on selection changes when selection level is not 0", () => {
                const tree = shallow(<PresentationTree
                    dataProvider={dataProviderMock.object}
                    selectionHandler={selectionHandlerMock.object}
                />);
                const s = spy.on(tree.instance(), Tree.prototype.render.name);
                triggerSelectionChange(1);
                expect(s).to.not.be.called();
            });

        });

    });

});
