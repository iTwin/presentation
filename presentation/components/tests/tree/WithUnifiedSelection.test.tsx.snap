// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree withUnifiedSelection renders correctly 1`] = `
<Tree
  dataProvider={null}
  onNodesDeselected={[Function]}
  onNodesSelected={[Function]}
  selectedNodes={[Function]}
  selectionMode={1}
/>
`;
